File: packages/zeal-go-sdk/subscription.go
package zeal

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// SubscriptionOptions configures webhook subscriptions
type SubscriptionOptions struct {
	Port             int               `json:"port"`
	Host             string            `json:"host"`
	Path             string            `json:"path"`
	HTTPS            bool              `json:"https"`
	Key              string            `json:"key"`
	Cert             string            `json:"cert"`
	AutoRegister     bool              `json:"autoRegister"`
	Namespace        string            `json:"namespace"`
	Events           []string          `json:"events"`
	BufferSize       int               `json:"bufferSize"`
	Headers          map[string]string `json:"headers"`
	VerifySignature  bool              `json:"verifySignature"`
	SecretKey        string            `json:"secretKey"`
}

// DefaultSubscriptionOptions returns default subscription options
func DefaultSubscriptionOptions() SubscriptionOptions {
	return SubscriptionOptions{
		Port:            3001,
		Host:            "0.0.0.0",
		Path:            "/webhooks",
		HTTPS:           false,
		AutoRegister:    true,
		Namespace:       "default",
		Events:          []string{"*"},
		BufferSize:      1000,
		VerifySignature: false,
	}
}

// WebhookDelivery represents a webhook delivery payload
type WebhookDelivery struct {
	WebhookID string                   `json:"webhook_id"`
	Events    []map[string]interface{} `json:"events"`
	Metadata  WebhookMetadata          `json:"metadata"`
}

// WebhookMetadata contains webhook delivery metadata
type WebhookMetadata struct {
	Namespace  string `json:"namespace"`
	DeliveryID string `json:"delivery_id"`
	Timestamp  string `json:"timestamp"`
}

// WebhookEventCallback is called for each webhook event
type WebhookEventCallback func(event map[string]interface{}) error

// WebhookDeliveryCallback is called for each webhook delivery
type WebhookDeliveryCallback func(delivery WebhookDelivery) error

// WebhookErrorCallback is called when an error occurs
type WebhookErrorCallback func(error) error

// WebhookObservable provides an observable interface for webhook events
type WebhookObservable struct {
	eventChan    chan map[string]interface{}
	errorChan    chan error
	completeChan chan struct{}
	subscription *WebhookSubscriptionManager
}

// Subscribe subscribes to webhook events with callbacks
func (wo *WebhookObservable) Subscribe(
	next WebhookEventCallback,
	errorHandler WebhookErrorCallback,
	complete func(),
) func() {
	ctx, cancel := context.WithCancel(context.Background())
	
	go func() {
		for {
			select {
			case event := <-wo.eventChan:
				if err := next(event); err != nil && errorHandler != nil {
					errorHandler(err)
				}
			case err := <-wo.errorChan:
				if errorHandler != nil {
					errorHandler(err)
				}
			case <-wo.completeChan:
				if complete != nil {
					complete()
				}
				return
			case <-ctx.Done():
				return
			}
		}
	}()
	
	return cancel
}

// Filter creates a filtered observable
func (wo *WebhookObservable) Filter(predicate func(map[string]interface{}) bool) *WebhookObservable {
	filtered := &WebhookObservable{
		eventChan:    make(chan map[string]interface{}, wo.subscription.options.BufferSize),
		errorChan:    make(chan error, 10),
		completeChan: make(chan struct{}),
		subscription: wo.subscription,
	}
	
	go func() {
		for {
			select {
			case event := <-wo.eventChan:
				if predicate(event) {
					filtered.eventChan <- event
				}
			case err := <-wo.errorChan:
				filtered.errorChan <- err
			case <-wo.completeChan:
				close(filtered.completeChan)
				return
			}
		}
	}()
	
	return filtered
}

// WebhookSubscriptionManager manages webhook subscriptions
type WebhookSubscriptionManager struct {
	webhooksAPI       *WebhooksAPI
	options           SubscriptionOptions
	server            *http.Server
	eventCallbacks    []WebhookEventCallback
	deliveryCallbacks []WebhookDeliveryCallback
	errorCallbacks    []WebhookErrorCallback
	webhookID         string
	isRunning         bool
	observable        *WebhookObservable
	mu                sync.RWMutex
}

// NewWebhookSubscription creates a new webhook subscription
func NewWebhookSubscription(webhooksAPI *WebhooksAPI, options *SubscriptionOptions) *WebhookSubscriptionManager {
	opts := DefaultSubscriptionOptions()
	if options != nil {
		if options.Port != 0 {
			opts.Port = options.Port
		}
		if options.Host != "" {
			opts.Host = options.Host
		}
		if options.Path != "" {
			opts.Path = options.Path
		}
		opts.HTTPS = options.HTTPS
		if options.Key != "" {
			opts.Key = options.Key
		}
		if options.Cert != "" {
			opts.Cert = options.Cert
		}
		opts.AutoRegister = options.AutoRegister
		if options.Namespace != "" {
			opts.Namespace = options.Namespace
		}
		if len(options.Events) > 0 {
			opts.Events = options.Events
		}
		if options.BufferSize > 0 {
			opts.BufferSize = options.BufferSize
		}
		if options.Headers != nil {
			opts.Headers = options.Headers
		}
		opts.VerifySignature = options.VerifySignature
		if options.SecretKey != "" {
			opts.SecretKey = options.SecretKey
		}
	}
	
	ws := &WebhookSubscriptionManager{
		webhooksAPI: webhooksAPI,
		options:     opts,
		observable: &WebhookObservable{
			eventChan:    make(chan map[string]interface{}, opts.BufferSize),
			errorChan:    make(chan error, 10),
			completeChan: make(chan struct{}),
		},
	}
	ws.observable.subscription = ws
	
	return ws
}

// OnEvent subscribes to webhook events with a callback
func (ws *WebhookSubscriptionManager) OnEvent(callback WebhookEventCallback) func() {
	ws.mu.Lock()
	ws.eventCallbacks = append(ws.eventCallbacks, callback)
	index := len(ws.eventCallbacks) - 1
	ws.mu.Unlock()
	
	return func() {
		ws.mu.Lock()
		defer ws.mu.Unlock()
		if index < len(ws.eventCallbacks) {
			ws.eventCallbacks = append(ws.eventCallbacks[:index], ws.eventCallbacks[index+1:]...)
		}
	}
}

// OnDelivery subscribes to webhook deliveries with a callback
func (ws *WebhookSubscriptionManager) OnDelivery(callback WebhookDeliveryCallback) func() {
	ws.mu.Lock()
	ws.deliveryCallbacks = append(ws.deliveryCallbacks, callback)
	index := len(ws.deliveryCallbacks) - 1
	ws.mu.Unlock()
	
	return func() {
		ws.mu.Lock()
		defer ws.mu.Unlock()
		if index < len(ws.deliveryCallbacks) {
			ws.deliveryCallbacks = append(ws.deliveryCallbacks[:index], ws.deliveryCallbacks[index+1:]...)
		}
	}
}

// OnError subscribes to errors with a callback
func (ws *WebhookSubscriptionManager) OnError(callback WebhookErrorCallback) func() {
	ws.mu.Lock()
	ws.errorCallbacks = append(ws.errorCallbacks, callback)
	index := len(ws.errorCallbacks) - 1
	ws.mu.Unlock()
	
	return func() {
		ws.mu.Lock()
		defer ws.mu.Unlock()
		if index < len(ws.errorCallbacks) {
			ws.errorCallbacks = append(ws.errorCallbacks[:index], ws.errorCallbacks[index+1:]...)
		}
	}
}

// AsObservable returns the observable interface
func (ws *WebhookSubscriptionManager) AsObservable() *WebhookObservable {
	return ws.observable
}

// Start starts the webhook server
func (ws *WebhookSubscriptionManager) Start() error {
	ws.mu.Lock()
	defer ws.mu.Unlock()
	
	if ws.isRunning {
		return fmt.Errorf("webhook subscription is already running")
	}
	
	mux := http.NewServeMux()
	mux.HandleFunc(ws.options.Path, ws.webhookHandler)
	
	addr := fmt.Sprintf("%s:%d", ws.options.Host, ws.options.Port)
	ws.server = &http.Server{
		Addr:    addr,
		Handler: mux,
	}
	
	var err error
	if ws.options.HTTPS && ws.options.Key != "" && ws.options.Cert != "" {
		go func() {
			err = ws.server.ListenAndServeTLS(ws.options.Cert, ws.options.Key)
		}()
	} else {
		go func() {
			err = ws.server.ListenAndServe()
		}()
	}
	
	if err != nil && err != http.ErrServerClosed {
		return fmt.Errorf("failed to start webhook server: %w", err)
	}
	
	// Give the server a moment to start
	time.Sleep(100 * time.Millisecond)
	
	ws.isRunning = true
	fmt.Printf("Webhook server listening on %s%s\n", addr, ws.options.Path)
	
	// Auto-register webhook if enabled
	if ws.options.AutoRegister {
		go func() {
			// Small delay to ensure server is fully started
			time.Sleep(200 * time.Millisecond)
			if err := ws.Register(); err != nil {
				ws.emitError(fmt.Errorf("failed to auto-register webhook: %w", err))
			}
		}()
	}
	
	return nil
}

// Stop stops the webhook server
func (ws *WebhookSubscriptionManager) Stop() error {
	ws.mu.Lock()
	defer ws.mu.Unlock()
	
	if !ws.isRunning {
		return nil
	}
	
	// Unregister webhook if it was registered
	if ws.webhookID != "" {
		if _, err := ws.webhooksAPI.Delete(context.Background(), ws.webhookID); err != nil {
			fmt.Printf("Failed to unregister webhook %s: %v\n", ws.webhookID, err)
		} else {
			fmt.Printf("Unregistered webhook %s\n", ws.webhookID)
		}
		ws.webhookID = ""
	}
	
	// Stop the server
	if ws.server != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		
		if err := ws.server.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown webhook server: %w", err)
		}
		fmt.Println("Webhook server stopped")
	}
	
	ws.isRunning = false
	close(ws.observable.completeChan)
	
	return nil
}

// Register registers the webhook with Zeal
func (ws *WebhookSubscriptionManager) Register() error {
	if !ws.isRunning {
		return fmt.Errorf("webhook server must be running before registration")
	}
	
	// Determine the public URL for the webhook
	protocol := "http"
	if ws.options.HTTPS {
		protocol = "https"
	}
	
	host := ws.options.Host
	if host == "0.0.0.0" {
		host = "localhost"
	}
	
	webhookURL := fmt.Sprintf("%s://%s:%d%s", protocol, host, ws.options.Port, ws.options.Path)
	
	// Register with Zeal
	req := CreateWebhookRequest{
		URL:     webhookURL,
		Events:  ws.options.Events,
		Headers: ws.options.Headers,
	}
	
	result, err := ws.webhooksAPI.Create(context.Background(), req)
	if err != nil {
		return fmt.Errorf("failed to register webhook: %w", err)
	}
	
	ws.webhookID = result.Subscription.ID
	fmt.Printf("Registered webhook %s at %s\n", ws.webhookID, webhookURL)
	
	return nil
}

// IsRunning returns whether the subscription is running
func (ws *WebhookSubscriptionManager) IsRunning() bool {
	ws.mu.RLock()
	defer ws.mu.RUnlock()
	return ws.isRunning
}

// WebhookID returns the current webhook ID if registered
func (ws *WebhookSubscriptionManager) WebhookID() string {
	ws.mu.RLock()
	defer ws.mu.RUnlock()
	return ws.webhookID
}

// OnEventType subscribes to specific event types
func (ws *WebhookSubscriptionManager) OnEventType(eventTypes []string, callback WebhookEventCallback) func() {
	eventTypeSet := make(map[string]bool)
	for _, eventType := range eventTypes {
		eventTypeSet[eventType] = true
	}
	
	filteredCallback := func(event map[string]interface{}) error {
		if eventTypeStr, ok := event["type"].(string); ok {
			if eventTypeSet[eventTypeStr] {
				return callback(event)
			}
		}
		return nil
	}
	
	return ws.OnEvent(filteredCallback)
}

// OnEventSource subscribes to events from specific sources
func (ws *WebhookSubscriptionManager) OnEventSource(sources []string, callback WebhookEventCallback) func() {
	sourceSet := make(map[string]bool)
	for _, source := range sources {
		sourceSet[source] = true
	}
	
	filteredCallback := func(event map[string]interface{}) error {
		if workflowID, ok := event["workflowId"].(string); ok {
			if sourceSet[workflowID] {
				return callback(event)
			}
		}
		return nil
	}
	
	return ws.OnEvent(filteredCallback)
}

// FilterEvents creates a filtered observable
func (ws *WebhookSubscriptionManager) FilterEvents(predicate func(map[string]interface{}) bool) *WebhookObservable {
	return ws.observable.Filter(predicate)
}

func (ws *WebhookSubscriptionManager) webhookHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}
	
	// Read the request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Failed to read request body", http.StatusBadRequest)
		ws.emitError(fmt.Errorf("failed to read request body: %w", err))
		return
	}
	defer r.Body.Close()
	
	// Verify signature if enabled
	if ws.options.VerifySignature && ws.options.SecretKey != "" {
		signature := r.Header.Get("X-Zeal-Signature")
		if !ws.verifySignature(body, signature) {
			http.Error(w, "Invalid signature", http.StatusUnauthorized)
			ws.emitError(fmt.Errorf("invalid webhook signature"))
			return
		}
	}
	
	// Parse the delivery
	var delivery WebhookDelivery
	if err := json.Unmarshal(body, &delivery); err != nil {
		http.Error(w, "Invalid JSON", http.StatusBadRequest)
		ws.emitError(fmt.Errorf("failed to parse webhook delivery: %w", err))
		return
	}
	
	// Process the delivery
	go ws.processDelivery(delivery)
	
	// Send success response
	w.WriteHeader(http.StatusOK)
	w.Write([]byte("OK"))
}

func (ws *WebhookSubscriptionManager) processDelivery(delivery WebhookDelivery) {
	// Call delivery callbacks
	ws.mu.RLock()
	deliveryCallbacks := make([]WebhookDeliveryCallback, len(ws.deliveryCallbacks))
	copy(deliveryCallbacks, ws.deliveryCallbacks)
	ws.mu.RUnlock()
	
	for _, callback := range deliveryCallbacks {
		if err := callback(delivery); err != nil {
			ws.emitError(fmt.Errorf("delivery callback error: %w", err))
		}
	}
	
	// Process individual events
	for _, event := range delivery.Events {
		// Send to observable
		select {
		case ws.observable.eventChan <- event:
		default:
			// Channel is full, skip this event
			ws.emitError(fmt.Errorf("event channel is full, skipping event"))
		}
		
		// Call event callbacks
		ws.mu.RLock()
		eventCallbacks := make([]WebhookEventCallback, len(ws.eventCallbacks))
		copy(eventCallbacks, ws.eventCallbacks)
		ws.mu.RUnlock()
		
		for _, callback := range eventCallbacks {
			if err := callback(event); err != nil {
				ws.emitError(fmt.Errorf("event callback error: %w", err))
			}
		}
	}
}

func (ws *WebhookSubscriptionManager) emitError(err error) {
	// Call error callbacks
	ws.mu.RLock()
	errorCallbacks := make([]WebhookErrorCallback, len(ws.errorCallbacks))
	copy(errorCallbacks, ws.errorCallbacks)
	ws.mu.RUnlock()
	
	for _, callback := range errorCallbacks {
		if callbackErr := callback(err); callbackErr != nil {
			fmt.Printf("Error callback failed: %v\n", callbackErr)
		}
	}
	
	// Send to observable
	select {
	case ws.observable.errorChan <- err:
	default:
		// Error channel is full
		fmt.Printf("Error channel is full, error: %v\n", err)
	}
}

func (ws *WebhookSubscriptionManager) verifySignature(body []byte, signature string) bool {
	if ws.options.SecretKey == "" {
		return false
	}
	
	// Parse the signature (format: "sha256=...")
	if !strings.HasPrefix(signature, "sha256=") {
		return false
	}
	
	expectedSig := signature[7:] // Remove "sha256=" prefix
	
	// Calculate HMAC
	mac := hmac.New(sha256.New, []byte(ws.options.SecretKey))
	mac.Write(body)
	calculatedSig := hex.EncodeToString(mac.Sum(nil))
	
	// Compare signatures
	return hmac.Equal([]byte(expectedSig), []byte(calculatedSig))
}

// WebhookConfig represents webhook configuration for registration
type WebhookConfig struct {
	Namespace string            `json:"namespace"`
	URL       string            `json:"url"`
	Events    []string          `json:"events"`
	Headers   map[string]string `json:"headers"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

// WebhookRegistrationResult represents the result of webhook registration
type WebhookRegistrationResult struct {
	WebhookID string `json:"webhookId"`
	URL       string `json:"url"`
}

File: packages/zeal-python-sdk/zeal/subscription.py
"""Webhook subscription functionality for Zeal Python SDK."""

import asyncio
import hashlib
import hmac
import json
import signal
import ssl
from asyncio import AbstractEventLoop
from contextlib import asynccontextmanager
from dataclasses import dataclass, field
from typing import Any, AsyncGenerator, Callable, Dict, List, Optional, Set, Union

from aiohttp import web, ClientSession
from aiohttp.web import Request, Response
from pydantic import BaseModel, Field

from .events import ZipWebhookEvent, parse_zip_webhook_event
from .webhooks import WebhooksAPI


@dataclass
class SubscriptionOptions:
    """Options for configuring webhook subscriptions."""
    port: int = 3001
    host: str = "0.0.0.0"
    path: str = "/webhooks"
    https: bool = False
    ssl_context: Optional[ssl.SSLContext] = None
    auto_register: bool = True
    namespace: str = "default"
    events: List[str] = field(default_factory=lambda: ["*"])
    buffer_size: int = 1000
    headers: Optional[Dict[str, str]] = None
    verify_signature: bool = False
    secret_key: Optional[str] = None


class WebhookMetadata(BaseModel):
    """Webhook delivery metadata."""
    namespace: str
    delivery_id: str = Field(alias="deliveryId")
    timestamp: str
    
    class Config:
        populate_by_name = True


class WebhookDelivery(BaseModel):
    """Webhook delivery containing multiple events."""
    webhook_id: str = Field(alias="webhookId")
    events: List[Dict[str, Any]]
    metadata: WebhookMetadata
    
    class Config:
        populate_by_name = True


# Type aliases for callbacks
WebhookEventCallback = Callable[[ZipWebhookEvent], None]
AsyncWebhookEventCallback = Callable[[ZipWebhookEvent], Any]  # Can be sync or async
WebhookDeliveryCallback = Callable[[WebhookDelivery], None]
AsyncWebhookDeliveryCallback = Callable[[WebhookDelivery], Any]  # Can be sync or async
WebhookErrorCallback = Callable[[Exception], None]
AsyncWebhookErrorCallback = Callable[[Exception], Any]  # Can be sync or async


class WebhookObservable:
    """Observable implementation for webhook events."""
    
    def __init__(self, buffer_size: int = 1000):
        self._subscribers: List[Dict[str, Any]] = []
        self._buffer_size = buffer_size
        self._event_queue: asyncio.Queue = asyncio.Queue(maxsize=buffer_size)
        self._error_queue: asyncio.Queue = asyncio.Queue(maxsize=10)
        self._completed = False
    
    def subscribe(
        self,
        next_handler: AsyncWebhookEventCallback,
        error_handler: Optional[AsyncWebhookErrorCallback] = None,
        complete_handler: Optional[Callable[[], Any]] = None
    ) -> Callable[[], None]:
        """Subscribe to webhook events."""
        subscriber = {
            'next': next_handler,
            'error': error_handler,
            'complete': complete_handler
        }
        self._subscribers.append(subscriber)
        
        def unsubscribe():
            if subscriber in self._subscribers:
                self._subscribers.remove(subscriber)
        
        return unsubscribe
    
    def filter(self, predicate: Callable[[ZipWebhookEvent], bool]) -> 'WebhookObservable':
        """Create a filtered observable."""
        filtered = WebhookObservable(self._buffer_size)
        
        async def filter_handler(event: ZipWebhookEvent):
            if predicate(event):
                await filtered.emit(event)
        
        self.subscribe(
            filter_handler,
            lambda err: filtered.error(err),
            lambda: filtered.complete()
        )
        
        return filtered
    
    async def emit(self, event: ZipWebhookEvent):
        """Emit an event to all subscribers."""
        if self._completed:
            return
        
        # Send to subscribers
        for subscriber in self._subscribers.copy():
            try:
                handler = subscriber['next']
                if asyncio.iscoroutinefunction(handler):
                    await handler(event)
                else:
                    handler(event)
            except Exception as e:
                if subscriber.get('error'):
                    try:
                        error_handler = subscriber['error']
                        if asyncio.iscoroutinefunction(error_handler):
                            await error_handler(e)
                        else:
                            error_handler(e)
                    except Exception:
                        pass  # Ignore errors in error handlers
    
    def error(self, err: Exception):
        """Emit an error to all subscribers."""
        if self._completed:
            return
        
        for subscriber in self._subscribers.copy():
            if subscriber.get('error'):
                try:
                    error_handler = subscriber['error']
                    if asyncio.iscoroutinefunction(error_handler):
                        asyncio.create_task(error_handler(err))
                    else:
                        error_handler(err)
                except Exception:
                    pass  # Ignore errors in error handlers
    
    def complete(self):
        """Complete the observable."""
        if self._completed:
            return
        
        self._completed = True
        for subscriber in self._subscribers.copy():
            if subscriber.get('complete'):
                try:
                    complete_handler = subscriber['complete']
                    if asyncio.iscoroutinefunction(complete_handler):
                        asyncio.create_task(complete_handler())
                    else:
                        complete_handler()
                except Exception:
                    pass  # Ignore errors in complete handlers
        
        self._subscribers.clear()


class WebhookSubscription:
    """Webhook subscription manager for receiving webhook events."""
    
    def __init__(self, webhooks_api: WebhooksAPI, options: Optional[SubscriptionOptions] = None):
        self.webhooks_api = webhooks_api
        self.options = options or SubscriptionOptions()
        self._server: Optional[web.Application] = None
        self._server_runner: Optional[web.AppRunner] = None
        self._site: Optional[web.TCPSite] = None
        
        self._event_callbacks: List[AsyncWebhookEventCallback] = []
        self._delivery_callbacks: List[AsyncWebhookDeliveryCallback] = []
        self._error_callbacks: List[AsyncWebhookErrorCallback] = []
        
        self._webhook_id: Optional[str] = None
        self._is_running = False
        self._observable = WebhookObservable(self.options.buffer_size)
        
        # Graceful shutdown handling
        self._shutdown_event = asyncio.Event()
        self._setup_signal_handlers()
    
    def _setup_signal_handlers(self):
        """Setup signal handlers for graceful shutdown."""
        try:
            for sig in (signal.SIGTERM, signal.SIGINT):
                signal.signal(sig, lambda s, f: asyncio.create_task(self.stop()))
        except (OSError, RuntimeError):
            # Signal handling might not work in all environments
            pass
    
    def on_event(self, callback: AsyncWebhookEventCallback) -> Callable[[], None]:
        """Subscribe to webhook events with a callback."""
        self._event_callbacks.append(callback)
        
        def unsubscribe():
            if callback in self._event_callbacks:
                self._event_callbacks.remove(callback)
        
        return unsubscribe
    
    def on_delivery(self, callback: AsyncWebhookDeliveryCallback) -> Callable[[], None]:
        """Subscribe to webhook deliveries with a callback."""
        self._delivery_callbacks.append(callback)
        
        def unsubscribe():
            if callback in self._delivery_callbacks:
                self._delivery_callbacks.remove(callback)
        
        return unsubscribe
    
    def on_error(self, callback: AsyncWebhookErrorCallback) -> Callable[[], None]:
        """Subscribe to errors with a callback."""
        self._error_callbacks.append(callback)
        
        def unsubscribe():
            if callback in self._error_callbacks:
                self._error_callbacks.remove(callback)
        
        return unsubscribe
    
    def as_observable(self) -> WebhookObservable:
        """Get the observable interface."""
        return self._observable
    
    async def start(self) -> None:
        """Start the webhook server."""
        if self._is_running:
            raise RuntimeError("Webhook subscription is already running")
        
        # Create the web application
        self._server = web.Application()
        self._server.router.add_post(self.options.path, self._webhook_handler)
        
        # Setup and start the server
        self._server_runner = web.AppRunner(self._server)
        await self._server_runner.setup()
        
        self._site = web.TCPSite(
            self._server_runner,
            host=self.options.host,
            port=self.options.port,
            ssl_context=self.options.ssl_context if self.options.https else None
        )
        
        await self._site.start()
        
        protocol = "https" if self.options.https else "http"
        print(f"Webhook server listening on {protocol}://{self.options.host}:{self.options.port}{self.options.path}")
        
        self._is_running = True
        
        # Auto-register webhook if enabled
        if self.options.auto_register:
            await asyncio.sleep(0.1)  # Small delay to ensure server is ready
            await self.register()
    
    async def stop(self) -> None:
        """Stop the webhook server."""
        if not self._is_running:
            return
        
        # Unregister webhook if it was registered
        if self._webhook_id:
            try:
                await self.webhooks_api.delete(self._webhook_id)
                print(f"Unregistered webhook {self._webhook_id}")
            except Exception as e:
                print(f"Failed to unregister webhook {self._webhook_id}: {e}")
            finally:
                self._webhook_id = None
        
        # Stop the server
        if self._site:
            await self._site.stop()
            self._site = None
        
        if self._server_runner:
            await self._server_runner.cleanup()
            self._server_runner = None
        
        self._server = None
        self._is_running = False
        
        # Complete the observable
        self._observable.complete()
        
        print("Webhook server stopped")
    
    async def register(self) -> None:
        """Register the webhook with Zeal."""
        if not self._is_running:
            raise RuntimeError("Webhook server must be running before registration")
        
        # Determine the public URL for the webhook
        protocol = "https" if self.options.https else "http"
        host = "localhost" if self.options.host == "0.0.0.0" else self.options.host
        webhook_url = f"{protocol}://{host}:{self.options.port}{self.options.path}"
        
        # Register with Zeal
        from .types import CreateWebhookRequest
        
        request = CreateWebhookRequest(
            url=webhook_url,
            events=self.options.events,
            headers=self.options.headers or {}
        )
        
        result = await self.webhooks_api.create(request)
        self._webhook_id = result.subscription.id
        print(f"Registered webhook {self._webhook_id} at {webhook_url}")
    
    def is_running(self) -> bool:
        """Check if the subscription is running."""
        return self._is_running
    
    def webhook_id(self) -> Optional[str]:
        """Get the current webhook ID if registered."""
        return self._webhook_id
    
    def on_event_type(
        self,
        event_types: Union[str, List[str]],
        callback: AsyncWebhookEventCallback
    ) -> Callable[[], None]:
        """Subscribe to specific event types."""
        types_set = set(event_types if isinstance(event_types, list) else [event_types])
        
        async def filtered_callback(event: ZipWebhookEvent):
            if event.type in types_set:
                if asyncio.iscoroutinefunction(callback):
                    await callback(event)
                else:
                    callback(event)
        
        return self.on_event(filtered_callback)
    
    def on_event_source(
        self,
        sources: Union[str, List[str]],
        callback: AsyncWebhookEventCallback
    ) -> Callable[[], None]:
        """Subscribe to events from specific sources."""
        sources_set = set(sources if isinstance(sources, list) else [sources])
        
        async def filtered_callback(event: ZipWebhookEvent):
            if hasattr(event, 'workflow_id') and event.workflow_id in sources_set:
                if asyncio.iscoroutinefunction(callback):
                    await callback(event)
                else:
                    callback(event)
        
        return self.on_event(filtered_callback)
    
    def filter_events(self, predicate: Callable[[ZipWebhookEvent], bool]) -> WebhookObservable:
        """Create a filtered observable."""
        return self._observable.filter(predicate)
    
    async def _webhook_handler(self, request: Request) -> Response:
        """Handle incoming webhook requests."""
        try:
            # Read request body
            body = await request.read()
            body_text = body.decode('utf-8')
            
            # Verify signature if enabled
            if self.options.verify_signature and self.options.secret_key:
                signature = request.headers.get('X-Zeal-Signature', '')
                if not self._verify_signature(body, signature):
                    await self._emit_error(Exception("Invalid webhook signature"))
                    return Response(status=401, text="Invalid signature")
            
            # Parse the delivery
            delivery_data = json.loads(body_text)
            delivery = WebhookDelivery(**delivery_data)
            
            # Process the delivery in background
            asyncio.create_task(self._process_delivery(delivery))
            
            return Response(status=200, text="OK")
        
        except Exception as e:
            await self._emit_error(e)
            return Response(status=500, text="Internal server error")
    
    async def _process_delivery(self, delivery: WebhookDelivery):
        """Process a webhook delivery."""
        try:
            # Call delivery callbacks
            for callback in self._delivery_callbacks.copy():
                try:
                    if asyncio.iscoroutinefunction(callback):
                        await callback(delivery)
                    else:
                        callback(delivery)
                except Exception as e:
                    await self._emit_error(e)
            
            # Process individual events
            for event_data in delivery.events:
                try:
                    # Parse the event
                    event = parse_zip_webhook_event(event_data)
                    
                    # Emit to observable
                    await self._observable.emit(event)
                    
                    # Call event callbacks
                    for callback in self._event_callbacks.copy():
                        try:
                            if asyncio.iscoroutinefunction(callback):
                                await callback(event)
                            else:
                                callback(event)
                        except Exception as e:
                            await self._emit_error(e)
                
                except Exception as e:
                    await self._emit_error(Exception(f"Failed to parse event: {e}"))
        
        except Exception as e:
            await self._emit_error(e)
    
    async def _emit_error(self, error: Exception):
        """Emit an error to all error callbacks and the observable."""
        # Call error callbacks
        for callback in self._error_callbacks.copy():
            try:
                if asyncio.iscoroutinefunction(callback):
                    await callback(error)
                else:
                    callback(error)
            except Exception:
                pass  # Ignore errors in error callbacks
        
        # Emit to observable
        self._observable.error(error)
    
    def _verify_signature(self, body: bytes, signature: str) -> bool:
        """Verify webhook signature."""
        if not self.options.secret_key:
            return False
        
        # Parse the signature (format: "sha256=...")
        if not signature.startswith("sha256="):
            return False
        
        expected_sig = signature[7:]  # Remove "sha256=" prefix
        
        # Calculate HMAC
        calculated_sig = hmac.new(
            self.options.secret_key.encode('utf-8'),
            body,
            hashlib.sha256
        ).hexdigest()
        
        # Compare signatures
        return hmac.compare_digest(expected_sig, calculated_sig)
    
    async def __aenter__(self):
        """Async context manager entry."""
        await self.start()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.stop()


# Convenience function for creating subscriptions
def create_webhook_subscription(
    webhooks_api: WebhooksAPI,
    options: Optional[SubscriptionOptions] = None
) -> WebhookSubscription:
    """Create a new webhook subscription."""
    return WebhookSubscription(webhooks_api, options)


# Async context manager for webhook subscriptions
@asynccontextmanager
async def webhook_subscription(
    webhooks_api: WebhooksAPI,
    options: Optional[SubscriptionOptions] = None
) -> AsyncGenerator[WebhookSubscription, None]:
    """Async context manager for webhook subscriptions."""
    subscription = WebhookSubscription(webhooks_api, options)
    try:
        await subscription.start()
        yield subscription
    finally:
        await subscription.stop()